core/src/main/java/org/elasticsearch/tasks/TaskResultsService.java

```diff
@@ -0,0 +1,158 @@
+            createIndexRequest.mapping(TASK_RESULT_TYPE, taskResultIndexMapping());
+            createIndexRequest.cause("auto(task api)");
+
+            createIndexAction.execute(null, createIndexRequest, new ActionListener<CreateIndexResponse>() {
+                @Override
+                public void onResponse(CreateIndexResponse result) {
+                    doPersist(taskResult, listener);
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    if (ExceptionsHelper.unwrapCause(e) instanceof IndexAlreadyExistsException) {
+                        // we have the index, do it
+                        try {
+                            doPersist(taskResult, listener);
+                        } catch (Throwable e1) {
+                            listener.onFailure(e1);
+                        }
+                    } else {
+                        listener.onFailure(e);
+                    }
+                }
+            });
+        } else {
+            IndexMetaData metaData = state.getMetaData().index(TASK_RESULT_INDEX);
+            if (metaData.getMappings().containsKey(TASK_RESULT_TYPE) == false) {
+                // The index already exists but doesn't have our mapping
+                client.admin().indices().preparePutMapping(TASK_RESULT_INDEX).setType(TASK_RESULT_TYPE).setSource(taskResultIndexMapping())
+                    .execute(new ActionListener<PutMappingResponse>() {
+                                 @Override
+                                 public void onResponse(PutMappingResponse putMappingResponse) {
+                                     doPersist(taskResult, listener);
+                                 }
+
+                                 @Override
+                                 public void onFailure(Throwable e) {
+                                     listener.onFailure(e);
+                                 }
+                             }
+                    );
+            } else {
+                doPersist(taskResult, listener);
+            }
+        }
+    }
+
+
+    private void doPersist(TaskResult taskResult, ActionListener<Void> listener) {
+        client.prepareIndex(TASK_RESULT_INDEX, TASK_RESULT_TYPE, taskResult.getTaskId().toString()).setSource(taskResult.getResult())
+            .execute(new ActionListener<IndexResponse>() {
+                @Override
+                public void onResponse(IndexResponse indexResponse) {
+                    listener.onResponse(null);
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    listener.onFailure(e);
+                }
+            });
+
+    }
+
+    private Settings taskResultIndexSettings() {
+        return Settings.builder()
+            .put(IndexMetaData.INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), 1)
+            .put(IndexMetaData.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), 1)
```

Comment:
`"auto_expand_replicas": 2` instead? That way you don't end up with a yellow cluster on one node and you end up with a fairly large amount of paranoia if you have three nodes? I'm ambivalent for the 2 vs 1 replica thing but I like `auto_expand_replicas` here.